# Исследование надежности заемщиков


## Описание проекта

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.
Описание данных
- children — количество детей в семье
- days_employed — общий трудовой стаж в днях
- dob_years — возраст клиента в годах
- education — уровень образования клиента
- education_id — идентификатор уровня образования
- family_status — семейное положение
- family_status_id — идентификатор семейного положения
- gender — пол клиента
- income_type — тип занятости
- debt — имел ли задолженность по возврату кредитов
- total_income — ежемесячный доход
- purpose — цель получения кредита

Импортирую данные и выведем на экран.

In [2]:
import pandas as pd

data = pd.read_csv('/datasets/data.csv')


FileNotFoundError: [Errno 2] No such file or directory: '/datasets/data.csv'

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Выведу информацию о датафрейме с помощью метода `info()`.**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Выведу количество пропущенных значений для каждого столбца**

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Заполню пропуски в столбце `total_income` медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Обработаю значения в столбце `days_employed`: заменим все отрицательные значения положительными с помощью метода `abs()`.**

In [ ]:
data['days_employed'] = data['days_employed'].abs()

**Для каждого типа занятости выведу медианное значение трудового стажа `days_employed` в днях.**

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьлю их как есть.

**Выведу перечень уникальных значений столбца `children`.**

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**В столбце `children` есть два аномальных значения. Удалю строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Ещё раз выведу перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

Заполню пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Убедюсь, что все пропуски заполнены. Проверю себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Заменю вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [ ]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Обработаю неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведу их к нижнему регистру.**

In [ ]:
data['education'] = data['education'].str.lower()

**Выведу на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалю их.**

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

### Категоризация данных

**На основании диапазонов, указанных ниже, создам в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Выведу на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Создам функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` появиться строка `'операции с автомобилем'`.**

**Использую собственную функцию с именем `categorize_purpose()` и метод `apply()`.**

In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data.head(30)

children  days_employed  dob_years            education  education_id  \
0          1    8437.673028         42               высшее             0   
1          1    4024.803754         36              среднее             1   
2          0    5623.422610         33              среднее             1   
3          3    4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0     926.185831         27               высшее             0   
6          0    2879.202052         43               высшее             0   
7          0     152.779569         50              среднее             1   
8          2    6929.865299         35               высшее             0   
9          0    2188.756445         41              среднее             1   
10         2    4171.483647         36               высшее             0   
11         0     792.701887         40              среднее             1   
12         0  365286.622650         65              среднее             1   
13         0    1846.641941         54  неоконченное высшее             2   
14         0    1844.956182         56               высшее             0   
15         1     972.364419         26              среднее             1   
16         0    1719.934226         35              среднее             1   
17         0    2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0   10038.818549         48              среднее             1   
20         1    1311.604166         36              среднее             1   
21         1     253.685166         33              среднее             1   
22         1    1766.644138         24              среднее             1   
23         0     272.981385         21               высшее             0   
24         1  338551.952911         57              среднее             1   
25         0  363548.489348         67              среднее             1   
26         0    2689.137274         41              среднее             1   
27         0     529.191635         28               высшее             0   
28         1     717.274324         26               высшее             0   
29         0  365286.622650         63              среднее             1   

            family_status  family_status_id gender  income_type  debt  \
0         женат / замужем                 0      F    сотрудник     0   
1         женат / замужем                 0      F    сотрудник     0   
2         женат / замужем                 0      M    сотрудник     0   
3         женат / замужем                 0      M    сотрудник     0   
4        гражданский брак                 1      F    пенсионер     0   
5        гражданский брак                 1      M    компаньон     0   
6         женат / замужем                 0      F    компаньон     0   
7         женат / замужем                 0      M    сотрудник     0   
8        гражданский брак                 1      F    сотрудник     0   
9         женат / замужем                 0      M    сотрудник     0   
10        женат / замужем                 0      M    компаньон     0   
11        женат / замужем                 0      F    сотрудник     0   
12       гражданский брак                 1      M    пенсионер     0   
13        женат / замужем                 0      F    сотрудник     0   
14       гражданский брак                 1      F    компаньон     1   
15        женат / замужем                 0      F    сотрудник     0   
16        женат / замужем                 0      F    сотрудник     0   
17       гражданский брак                 1      M    сотрудник     0   
18         вдовец / вдова                 2      F    пенсионер     0   
19              в разводе                 3      F    сотрудник     0   
20        женат / замужем                 0      M    сотрудник     0   
21       г

### Шаг 3. Исследование данных

#### 3.1 Рассмотрю, есть ли зависимость между количеством детей и возвратом кредита в срок?

In [ ]:
#создам функцию для сортировки клиетов по количеству детей.
#Объединю людей, имеющих 3 и более ребенка в категорию "больше 2х детей", так как от 3 детей семья считается многодетной.

def having_children(children):
    if children == 0:
        return 'нет детей'
    if children == 1:
        return 'один ребенок'
    if children == 2:
        return 'два ребенка'
    return 'больше двух детей'

data['having children'] = data['children'].apply(having_children)

#создам сводную таблицу на основе полученных данных, в которую объединю показатели"наличие детей" и возврат кредита в срок.

data_pivot = data.pivot_table(index = 'having children', values = 'debt',aggfunc=['sum', 'count',])


#посчитаю процентное отношение количества должников и общего количества людей с детьми
data_pivot['percentage'] = (data_pivot['sum']/data_pivot['count'])*100
data_pivot

NameError: name 'data' is not defined

**Вывод:** Доля количества должников с детьми чуть больше чем без оных, однако, замечу, что у людей с тремя и более детьми процент возникновения долга меньше, чем у людей с одним ребенком или двумя. Это может означать, что, если у людей есть доходы содержать большое количество детей, то и кредит они будут выплачивать без задержек. Стоит также заметить большой разброс величин по катериям: количество человек без детей - 14 091, а тех, у кого более двух детей - 380, что может повлиять на точность исследования.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [ ]:
data_pivot = data.pivot_table(index = 'family_status', values = 'debt',aggfunc=['sum', 'count',])
data_pivot['percentage'] = (data_pivot['sum']/data_pivot['count'])*100
data_pivot


sum  count percentage
                      debt   debt           
family_status                               
Не женат / не замужем  273   2796   9.763948
в разводе               84   1189   7.064760
вдовец / вдова          63    951   6.624606
гражданский брак       385   4134   9.313014
женат / замужем        927  12261   7.560558

**Вывод:** Наибольший процент невозврата кредита в срок наблюдается у не женатых/не замужних людей, а также у людей, состоящих в гражданском браке. Наименьший же процент долга у вдовцов/вдов и у людей в разводе.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [ ]:
data_pivot = data.pivot_table(index = 'total_income_category', values = 'debt',aggfunc=['sum', 'count',])
data_pivot['percentage'] = (data_pivot['sum']/data_pivot['count'])*100
data_pivot


sum  count percentage
                       debt   debt           
total_income_category                        
A                         2     25   8.000000
B                       354   5014   7.060231
C                      1353  15921   8.498210
D                        21    349   6.017192
E                         2     22   9.090909

**Вывод:** Как видно, наибольший процент должников у людей с категорией дохода "Е" (до 30 000), а наименьший с доходом "D" (до 50 000), однако согласно данным, разброс в категориях очень велик, категория Е - 22 человека, С - 15 921 человек, что не даёт объективной оценки.                                                                      

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [ ]:
data_pivot = data.pivot_table(index = 'purpose_category', values = 'debt',aggfunc=['sum', 'count',])
data_pivot['percentage'] = (data_pivot['sum']/data_pivot['count'])*100
data_pivot# Ваш код будет здесь. Вы можете создавать новые ячейки.

sum  count percentage
                         debt   debt           
purpose_category                               
операции с автомобилем    400   4279   9.347978
операции с недвижимостью  780  10751   7.255139
получение образования     369   3988   9.252758
проведение свадьбы        183   2313   7.911803

**Вывод:** Люди, берущие кредит на траты связанные с автомобилем и на образование, чаще не отдают его в срок: **9.3% и 9.2%** соответственно. Стоит заметить, что эти катеории примерно равны по количеству данных, а вот категория "операции с недвижимостью" почти в 2 раза больше по количеству данных, что помешает объективности исследования.

#### 3.5 Возможные причины появления пропусков в исходных данных.

В исходных данных пропуски мы обнаружили в столбцах "days_employed" и "total_income", что соответсвует отработанным дням и суммарному доходу клиентов. Пропуски здесь, вероятно, появились из-за того, что клиент не работал официально:
1. получал "серую" зарплату, которую невозможно отследить;
2. у него нет записей в трудовой книжке, поэтому невозможно понять всего сколько дней он отработал. 

#### 3.6 Заполнить пропуски медианным значением — лучшее решение для количественных переменных.

Медиана не подвержена влияниям аномальных знчений(по определению - это число, которое находится по середине выборки), в отличии от среднего арифметического, поэтому будет объективнее отражать данные.

### Шаг 4: общий вывод.

Проведя данное исследование, можно ответить на поставленный вопрос "влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок?".

**Да, семейное положение и количество детей вляет на выплату кредита в срок**, рассмотрю по-подробнее:

Во время работы над проектом, я:
- заменил пустые значения медианным
- сгруппировал клиентов по доходу, создав категории А, B, C, D, E и добавил новый стобец в таблицу
- сгруппировал клиентов по цели кредита, также добавив столбец в таблицу
- рассчитал нужные нам параметры

Влияние наличия детей на вероятность выплаты кредита в срок:
  - Нет детей ≈ 7.5%
  - 1 ребенок ≈ 9.2% 
  - 2 ребенка ≈ 9.5%
  - Более 2х детей ≈ 8.2%
  
Разница между максимальным значением и минимальным ≈ 2%

Теперь рассмотрим влияние семейного положения на вероятность просрочки платежа:

- не женат / не замужем ≈ 9.7%
- в разводе ≈ 7.1%
- вдовец / вдова ≈ 6.6%
- гражданский брак ≈ 9.3%
- женат / замужем ≈ 7.6%

Разница между максимальным значением и минимальным ≈ 2.6%

Зависимость между уровнем дохода и возвратом кредита в срок:

- A(до 30 000)	≈ 8%
- B(до 50 000)	≈ 7.1%
- C(до 200 000)	≈ 8.5%
- D(до 1 000 000) ≈ 6%
- E(>1 000 000)	≈ 9.1%

Разница между максимальным значением и минимальным ≈ 3.1%

Цели кредита и возврат в срок:

- операции с автомобилем ≈ 9.3 %
- операции с недвижимостью	≈ 7.3 %
- получение образования	≈ 9.3 %
- проведение свадьбы ≈ 8 %

Разница между максимальным значением и минимальным ≈ 2%

Исходя из получнных данных, чуть большее значение имеет семейный статус клиента.





